flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java

```diff
@@ -61,50 +64,103 @@
 			optionWithMetaInfo -> optionWithMetaInfo.field.getAnnotation(Documentation.CommonOption.class) != null);
 
+		assertDocumentedOptionsAreWellDefined(documentedOptions);
+		assertExistingOptionsAreWellDefined(existingOptions);
+
 		compareDocumentedAndExistingOptions(documentedOptions, existingOptions);
 	}
 
 	@Test
 	public void testFullReferenceCompleteness() throws IOException, ClassNotFoundException {
-		Map<String, DocumentedOption> documentedOptions = parseDocumentedOptions();
-		Map<String, ExistingOption> existingOptions = findExistingOptions(ignored -> true);
+		Map<String, List<DocumentedOption>> documentedOptions = parseDocumentedOptions();
+		Map<String, List<ExistingOption>> existingOptions = findExistingOptions(ignored -> true);
+
+		assertDocumentedOptionsAreWellDefined(documentedOptions);
+		assertExistingOptionsAreWellDefined(existingOptions);
 
 		compareDocumentedAndExistingOptions(documentedOptions, existingOptions);
 	}
 
-	private static void compareDocumentedAndExistingOptions(Map<String, DocumentedOption> documentedOptions, Map<String, ExistingOption> existingOptions) {
+	private static void assertDocumentedOptionsAreWellDefined(Map<String, List<DocumentedOption>> documentedOptions) {
+		assertOptionsAreWellDefined(documentedOptions, (option1, option2) -> {
+				// found a ConfigOption pair with the same key that aren't equal
+				// we fail here outright as this is not a documentation-completeness problem
+				if (!option1.defaultValue.equals(option2.defaultValue)) {
+					throw new AssertionError("Documentation contains distinct defaults for " +
+						option1.key + " in " + option1.containingFile + " and " + option2.containingFile + '.');
+				} else {
+					throw new AssertionError("Documentation contains distinct descriptions for " +
+						option1.key + " in " + option1.containingFile + " and " + option2.containingFile + '.');
+				}
+			});
+	}
+
+	private static void assertExistingOptionsAreWellDefined(Map<String, List<ExistingOption>> existingOptions) {
+		assertOptionsAreWellDefined(existingOptions, (option1, option2) -> {
+				// found a ConfigOption pair with the same key that aren't equal
+				// we fail here outright as this is not a documentation-completeness problem
+				if (!option1.defaultValue.equals(option2.defaultValue)) {
+					throw new AssertionError("Ambiguous option " + option1.key + " due to distinct default values (" + option1.defaultValue + " vs " + option2.defaultValue + ").");
+				} else {
+					throw new AssertionError("Ambiguous option " + option1.key + " due to distinct descriptions.");
+				}
+			});
+	}
+
+	private static <O> void assertOptionsAreWellDefined(Map<String, List<O>> allOptions, BiFunction<O, O, AssertionError> duplicateHandler) {
+		allOptions.entrySet().stream()
+			.filter(entry -> !WELL_DEFINED_WHITELIST.contains(entry.getKey()))
+			.map(Map.Entry::getValue)
+			.forEach(options -> options.stream().reduce((option1, option2) -> {
+				if (option1.equals(option2)) {
```

Comment:
I think we should better include `containingClass` to compare when calling `ExistingOption#equals` so that these two `ExistingOption`s are really identical.

Reply:
That would violate our definition of begin well-defined; from the javadocs:
`no 2 options exist for the same key with different descriptions/default values`

The organization into separate classes that we have is only for developer convenience and readability only; it does not (and must not) have any semantics attached to it. Hence the containing class doesn't matter.

The reason for this is simplicity; we don't have to worry about
* options clashing in their default value, which is difficult to document in a good way and can lead to subtle issues when de-duplicating options
* options clashing in their description, which is also difficult to document and usually leads to stale documentation at one place
* options clashing in their type, which may result in a component being unusable when used in conjunction with another
* users not being able to configure distinct values per option

Reply:
If some developers happen to create the same config key in different classes, I‘m afraid there exists some risk that user configure that key in `flink-conf.yaml` but taken effect by another unexpected class.

By the way, after introducing data type for config option, we did not include that field to compare in [duplication check](https://github.com/apache/flink/blob/df5eb214294f09a3cd7fcc04ab2617a321041e9d/flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java#L219), and this is certainly out of scope of this PR.

Reply:
> If some developers happen to create the same config key in different classes, I‘m afraid there exists some risk that user configure that key in flink-conf.yaml but taken effect by another unexpected class.

That's exactly what this check is preventing.

Reply:
Would you please kindly explain whether the scenario below is legal in Flink?  
We define two config option named as `a.b.c` with the same default value and description in different options files (e.g in `CheckpointingOptions` and `RocksDBOptions`). Current `ConfigOptionsDocsCompletenessITCase` would still pass successfully.

Reply:
That's legal. IIRC, when we introduced the generator we actually had an option that was mirroring another one. One could also argue that logically, if code-structure must not have semantics, duplication of code must be allowed.

However, it seems unlikely that 2 options are being created for which they
1. share a key
2. have an _exactly_ matching description (supposedly written in isolation from each other)
3. are semantically so aligned that the key/description are still meaningful
4. yet are semantically so different that the accidental configuration can cause problems.

Overall I think this is mostly a theoretical issue due to us usually providing a component context via the key; to take your example, all `RocksDBOptions` have `rocksdb` in their key, with the inverse rightfully being the case for the `CheckpointingOptions`.

Note that in any case the behavior in this regard isn't different to `master`.

Reply:
It is also not really the responsibility of this test to forbid this; which is apparent when you consider that whether you de-duplicate the option or not, the contents of the documentation remain the same.

Reply:
Thanks for your interpretation, I still reserve my judgment on this problem and I agree this should be out of scope of this PR since this problem has been existed for a while.